File: elixir/hamming/hamming.exs
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: non_neg_integer
  def hamming_distance(strand1, strand2) when strand1 == strand2 do
    {:ok, 0}
  end

  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2) do
    {:error, "Lists must be the same length"}
  end

  def hamming_distance(strand1, strand2) do
    dist = List.zip([strand1, strand2])
    |> Enum.reduce( 0, fn(pair, acc) ->
      acc + distance(pair)
    end)
    {:ok, dist}
  end

  defp distance({a,b}) when a === b, do: 0
  defp distance(_), do: 1

end

File: javascript/hamming/hamming.js
Hamming = (function() {
  function Hamming() {
  };

  function computeHammingDistance(strandA, strandB) {
    var len = strandA.length
    var distance = 0;
    for( var i = 0; i < len; ++i ) {
      if (strandA[i] !== strandB[i]) {
        distance = distance + 1;
      }
    }
    return distance;
  }

  Hamming.prototype.compute = function compute(strandA, strandB) {
    if (strandA.length != strandB.length) {
      throw new Error('DNA strands must be of equal length.')
    }
    return computeHammingDistance(strandA, strandB)
  };
  return Hamming
})();

module.exports = Hamming;
